docs/source/howto/data.rst

```diff
@@ -249,11 +247,693 @@ The node class has various methods to set these attributes, such as :py:`~aiida.
+
+Both APIs provide the same functionality - the appender method may be more suitable for everyday use in the ``verdi shell``, whereas the queryhelp method can be useful in scripting.
+
+.. _how-to:data:find:appender:
+
+The appender method
+-------------------
+
+Selecting entities
+^^^^^^^^^^^^^^^^^^
+
+Using the ``append()`` method of the ``QueryBuilder``, you can query for the entities you are interested in.
+Suppose you want to query for calculation job nodes in your database::
 
-`#3996`_
+    from aiida.orm.querybuilder import QueryBuilder
+    qb = QueryBuilder()       # Instantiating instance. One instance -> one query
+    qb.append(CalcJobNode)    # Setting first vertex of path
+
+If you are interested in instances of different classes, you can also pass a tuple, list or set of classes.
+However, they have to be of the same ORM-type (e.g. all have to be subclasses of ``Node``)::
+
+    qb = QueryBuilder()       # Instantiating instance. One instance -> one query
+    qb.append([CalcJobNode, WorkChainNode]) # Setting first vertice of path, either WorkChainNode or Job.
+
+.. note ::
+    Remember that :ref:`processes<topics:processes:concepts:types>` have both a run-time ``Process`` that steers them and a ``Node`` that stores their metadata in the database.
+    The QueryBuilder allows you to pass either the ``Node`` class (e.g. ``CalcJobNode``) or the ``Process`` class (e.g. ``CalcJob``), which will automatically apply the correct filters for the type of calculation.
+
+.. _how-to:data:find:appender:results:
+
+Retrieving results
+^^^^^^^^^^^^^^^^^^
+
+Once you have *appended* the entity you want to query for to the ``QueryBuilder``, the next question is how to get the results.
+There are several ways to obtain data from a query::
+
+    qb = QueryBuilder()                 # Instantiating instance
+    qb.append(CalcJobNode)              # Setting first vertice of path
+
+    first_row = qb.first()              # Returns a list (!)
+                                        # of the results of the first row
+
+    all_results_d = qb.dict()           # Returns all results as
+                                        # a list of dictionaries
+
+    all_results_l = qb.all()            # Returns a list of lists
+
+In case you are working with a large dataset, you can also return your query as a generator::
+
+    all_res_d_gen = qb.iterdict()       # Return a generator of dictionaries
+                                        # of all results
+    all_res_l_gen = qb.iterall()        # Returns a generator of lists
+
+This will retrieve the data in batches, and you can start working with the data before the query has completely finished.
+Be aware that when using generators, you should never commit (store) anything during the iteration.
+While the query is still running it can be compromised by new data in the database.
```

Comment:
I don't think the data is compromised. The way I understand the sentence is that the query can be compromised when storing new data during the query.

I'll add your suggested sentence as a warning for now, but leave this comment as unresolved.

Reply:
@sphuber Could you clarify on what exactly is at risk of being compromised here?

Reply:
I don't think that existing data can really be "compromised" or corrupted in so much as that updating a collection when iterating over it, is in general not advisable (note that comprised in the text refers to the query and not the data in the database). This also holds for Python itself, since the operation, fundamentally, is not well defined and *could* have unexpected behavior. Python even raises for this kind of behavior:
```
In [3]: d = {'a': 1, 'b': 2, 'c': 3}                                                                                                                                                                             

In [4]: for key in d: 
   ...:     if key == 'a': 
   ...:         d.pop('b') 
   ...:                                                                                                                                                                                                          
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-4-ce771193fc04> in <module>
----> 1 for key in d:
      2     if key == 'a':
      3         d.pop('b')
      4 

RuntimeError: dictionary changed size during iteration
```

To conclude: we could say that users should not create new entities when iterating over the relevant collection as the query might not make sense at some point. However, I think the chances of anyone doing this by accident is minimal, and at worst the query will crash. Don't think there will be much worse consequences but can not be certain

Reply:
So, perhaps it's not necessary to add a warning in the how-to, but rather move this discussion to the topics section?

Reply:
Yeah, I don't think this is enough of a problem to confuse users here.

Reply:
This could also be mentioned as part of the doc-string.